### ISSUE TITLE
[red-knot] Add a diagnostic for `raise` statements used with non-exceptions

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#15038]
Red-knot currently doesn't emit a diagnostic on this snippet:

```py
raise 42
```

But we should, since `42` is not a valid object to be used in a `raise` statement:

```pycon
>>> raise 42
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    raise 42
TypeError: exceptions must derive from BaseException
```

The rule is that the object used in a `raise` statement must be assignable to the type `BaseException | type[BaseException]`, since both of the following work:

```pycon
>>> raise ValueError
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    raise ValueError
ValueError
>>> raise ValueError()
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    raise ValueError()
ValueError
```

We need to add some extra logic to this branch here: https://github.com/astral-sh/ruff/blob/c9fdb1f5e3a4d0d60b4537741f2c9c19e2426038/crates/red_knot_python_semantic/src/types/infer.rs#L2193-L2201

I think this will probably need to be a new rule added to https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/src/types/diagnostic.rs ? I don't think it fits neatly into any other rule. The alternative might be to make the existing `INVALID_EXCEPTION_CAUGHT` rule broader (and rename it).

**Comment by MichaReiser** [comment]
> The alternative might be to make the existing INVALID_EXCEPTION_CAUGHT rule broader (and rename it).

That was also my first reaction. It's kind of the same problem.

**Comment by AlexWaygood** [comment]
> > The alternative might be to make the existing INVALID_EXCEPTION_CAUGHT rule broader (and rename it).
> 
> That was also my first reaction. It's kind of the same problem.

The same genre of problem, for sure. Though there's a subtle difference between the rules for raising exceptions and the rules for catching exceptions:
- If you're raising an exception, the type must be assignable to `BaseException | type[BaseException]`
- but if you're catching an exception the type must be assignable to `type[BaseException] | tuple[type[BaseException], ...]`

If we bundle them together into one rule, it'll be harder to explain the problem in the rule's documentation

**Comment by AlexWaygood** [comment]
I don't have a strong opinion, though. Any suggestions for what we might call the rule if we combined the two into one one rule?
